File: app/src/main/java/es/osw/cafelatte/presentation/di/base/component/ApplicationComponent.java
package es.osw.cafelatte.presentation.di.base.component;

import android.content.Context;
import com.google.gson.Gson;
import dagger.Component;
import es.osw.cafelatte.presentation.di.base.module.ApplicationModule;
import es.osw.cafelatte.presentation.di.base.module.DataBindingModule;
import es.osw.cafelatte.presentation.di.base.module.DataModule;
import es.osw.cafelatte.presentation.ui.CoffeeApp;
import javax.inject.Singleton;

@Singleton // Constraints this component to one-per-application or unscoped bindings.
@Component(modules = {
    ApplicationModule.class, DataBindingModule.class, DataModule.class
}) public interface ApplicationComponent extends ApplicatonDataDependencies {
  void inject(CoffeeApp app);

  Context context();

  Gson gson();
}

File: app/src/main/java/es/osw/cafelatte/presentation/di/base/component/ApplicationComponent.kt
package es.osw.cafelatte.presentation.di.base.component

import dagger.Component
import es.osw.cafelatte.presentation.di.base.module.ActivityModule
import es.osw.cafelatte.presentation.di.base.module.ApplicationModule
import es.osw.cafelatte.presentation.di.base.module.DBModule
import es.osw.cafelatte.presentation.di.base.module.DataBindingModule
import es.osw.cafelatte.presentation.ui.CoffeeApp
import es.osw.cafelatte.presentation.ui.creditcard.AddCreditCardActivity
import es.osw.cafelatte.presentation.ui.login.LoginActivity
import es.osw.cafelatte.presentation.ui.main.HomeActivity
import es.osw.cafelatte.presentation.ui.profile.ProfileActivity
import es.osw.cafelatte.presentation.ui.splash.SplashActivity
import javax.inject.Singleton

@Singleton
@Component(modules = arrayOf(ApplicationModule::class, DataBindingModule::class, DBModule::class))
interface ApplicationComponent : ApplicationDataDependencies {
    fun inject(app: CoffeeApp)

    fun homeComponent(activityModule: ActivityModule): HomeActivity.HomeComponent

    fun loginComponent(activityModule: ActivityModule): LoginActivity.LoginComponent

    fun addCreditCardComponent(activityModule: ActivityModule): AddCreditCardActivity.AddCreditCardComponent

    fun splashComponent(activityModule: ActivityModule): SplashActivity.SplashComponent

    fun profileComponent(activityModule: ActivityModule): ProfileActivity.ProfileComponent
}
